sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala

```diff
@@ -586,8 +586,9 @@ object FunctionRegistry {
         val params = Seq.fill(expressions.size)(classOf[Expression])
         val f = constructors.find(_.getParameterTypes.toSeq == params).getOrElse {
```

Comment:
but here we're still filtering by expressions..am I missing something?

Reply:
Two reasons:
1. It will throw exception at `CheckAnalysis` if parameter is a `DataType`. So we do not need handle it here:
```java
scala> spark.sql("select cast(int)")
org.apache.spark.sql.AnalysisException: cannot resolve '`int`' given input columns: []; line 1 pos 12;
'Project [unresolvedalias('cast('int), None)]
+- OneRowRelation

  at org.apache.spark.sql.catalyst.analysis.package$AnalysisErrorAt.failAnalysis(package.scala:42)
  at org.apache.spark.sql.catalyst.analysis.CheckAnalysis$$anonfun$$nestedInanonfun$checkAnalysis$1$2.applyOrElse(CheckAnalysis.scala:113)
  at org.apache.spark.sql.catalyst.analysis.CheckAnalysis$$anonfun$$nestedInanonfun$checkAnalysis$1$2.applyOrElse(CheckAnalysis.scala:110)
  at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$transformUp$2(TreeNode.scala:306)
```
2. It's hard to know a parameter is `DataType` or `Expression` here.

Reply:
So, I think the problem here is not that we are not considering the `DataType` args, but the problem IMHO is that we are not handling the case `validParametersCount == 0`. I think in that case we should throw something like: `Invalid arguments for function $name.`. Indeed, in the cast case, I'd argue that it is not so clear to me whether the arguments are 1 or 2, since it is not `cast(1, string)` but `cast(1 as string)`... So I don't think that it is correct to state that the number of arguments for cast is 2.